### PR TITLE
asp-format support on input[type=hidden] fields

### PR DESCRIPTION
**Pull request by Haplois**
Change log:
 - Added support for (will no longer ignore) `asp-format` when the chosen type is `'hidden'`.

Fixes https://github.com/aspnet/Mvc/issues/5094.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs [hunk 1]
```diff
@@ -189,20 +189,21 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
             TagBuilder tagBuilder;
             switch (inputType)
             {
-                case "checkbox":
-                    GenerateCheckBox(modelExplorer, output);
-                    return;
-
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: Curious: Did you reorder this case for a reason?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 2]
```diff
@@ -338,7 +339,36 @@ public virtual IHtmlContent GenerateAntiforgery(ViewContext viewContext)
                 isChecked: false,
                 setId: true,
                 isExplicitValue: true,
-                format: null,
+                format: format,
+                htmlAttributes: htmlAttributeDictionary);
+        }
+
+        /// <inheritdoc />
+        public virtual TagBuilder GenerateTextBox(
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: Please move this back where it was. It's hard to see what, if anything, changed in this method.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs [hunk 3]
```diff
@@ -957,6 +958,7 @@ protected virtual string GenerateDisplayText(ModelExplorer modelExplorer)
             ModelExplorer modelExplorer,
             string expression,
             object value,
+            string format,
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/HtmlHelper.cs]: Don't think `HtmlHelper.GenerateHidden()` needs to change other than passing `format: null` to `_htmlGenerator.GenerateHidden()`.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs [hunk 4]
```diff
@@ -157,6 +157,7 @@ public interface IHtmlGenerator
             ModelExplorer modelExplorer,
             string expression,
             object value,
+            string format,
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs]: @Eilon please reconfirm the addition of this parameter in the `public` but rarely-used-by-others `IHtmlGenerator` interface is :ok: for 1.1.0.

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs [hunk 5]
```diff
@@ -404,6 +404,7 @@ public static TheoryData MultiAttributeCheckBoxData
                     tagHelper.For.ModelExplorer,
                     tagHelper.For.Name,
                     model,      // value
+                    null,       // format
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs]: Please add a `format` parameter to this test as well as test cases where that value is non-`null`. Then use `format` here to confirm value is passed through to the generator as expected.

Also need a test in this class that confirms the `Format` property works correctly. Doesn't appear that's tested even when `type='text'` is chosen. Not your fault but since you're working in that area, it's a good time to fill that test gap.

**Comment by dougbu** [comment]
:watch: for a couple of change reversals and test additions.

Thanks very much for your work here @Haplois!

**Comment by Haplois** [comment]
No worries, I'll make those changes today.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs (hunk 4 above)
Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs]: It's a breaking change, so no.

Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs]: Can we come up with a clever way to not need this? And we can "fix" it all up in 2.0.0?

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs]: If we can't add a method to this interface (even though we don't expect another implementation), about all that comes to mind is duplicating the few lines of formatting code in the `<input/>` tag helper. @Haplois any other ideas?

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs]: (I mention "method" because that would leave callers using the old interface unaffected.)

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs (hunk 5 above)
Comment (Haplois) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/InputTagHelperTest.cs]: No worries, I'll write those tests.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs (hunk 4 above)
Comment (Haplois) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs]: The easiest way is comes into my mind feels like a hack. But, (for 1.1.0) release may be we should call `IHtmlGenerator.GenerateInput` instead of `IHtmlGenerator.GenerateHidden`. We'll duplicate 2 lines of serialization code that way, but nothing else will be affected.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs (hunk 1 above)
Comment (Haplois) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: Nope, in the https://github.com/aspnet/Mvc/pull/5109/commits/4b297d9bca77eea3ef526a65535f1d16835fb5d8 commit, I was removed `case "hidden"`, when I added it again I added it on top. (I thought `hidden`s are more common then `checkbox`es.)

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs (hunk 4 above)
Comment (Haplois) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs]: Nope, it appears `GenerateInput` is not a member of `IHtmlGenerator`. It's an `protected` member of `DefaultHtmlGenerator`.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs [hunk 6]
```diff
@@ -340,6 +334,29 @@ private TagBuilder GenerateTextBox(ModelExplorer modelExplorer, string inputType
                 htmlAttributes: htmlAttributes);
         }
 
+        private TagBuilder GenerateHidden(ModelExplorer modelExplorer)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: Definitely deserves a comment about mimicking `IHtmlGenerator.GenerateHidden()` with added `format` support. In particular, mention call to that protected `GenerateInput()` method in the `DefaultHtmlGenerator` treats `InputType.Text` and `InputType.Hidden` identically if the given HTML attributes already contain `type`.

**Comment by dougbu** [comment]
Please add tests mentioned [here](https://github.com/aspnet/Mvc/pull/5109/files/4b297d9bca77eea3ef526a65535f1d16835fb5d8#r73795321)

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs (hunk 6 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs]: BTW I like this approach given the 1.1.0 constraints :+1:

**Comment by dougbu** [comment]
@Haplois thanks very much for your submission. Please see #5205 for the tests and comments I added. Your commit remains and will be part of the final push to GitHub of course.
